File: two-pointers/0011_container_with_most_water.py
"""
Problem: Container With Most Water
LeetCode #: 11
Difficulty: Medium
Link: https://leetcode.com/problems/container-with-most-water/

Approach: Two pointers starting from ends. Move pointer pointing to shorter line inward.
Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        left, right = 0, len(height) - 1
        max_water = 0

        while left < right:
            h = min(height[left], height[right])
            w = right - left
            max_water = max(max_water, h * w)

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return max_water


# --- Test ---
if __name__ == "__main__":
    sol = Solution()
    print(sol.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))  # 49

File: two-pointers/0011_container_with_most_water.java
/*
 * Problem: Container With Most Water
 * LeetCode #: 11
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/container-with-most-water/
 *
 * Approach: Two pointers starting from ends. Move pointer pointing to shorter line inward.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

public class Solution {
    public int maxArea(int[] height) {
        int left = 0, right = height.length - 1;
        int maxWater = 0;

        while (left < right) {
            int h = Math.min(height[left], height[right]);
            int w = right - left;
            maxWater = Math.max(maxWater, h * w);

            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return maxWater;
    }

    // --- Test ---
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.maxArea(new int[]{1, 8, 6, 2, 5, 4, 8, 3, 7})); // 49
    }
}
